ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -261,30 +263,38 @@ public void setReplicaDisabledState(ReplicaId replicaId, boolean disable) {
           "HelixParticipant only works with the AmbryReplica implementation of ReplicaId");
     }
     synchronized (helixAdministrationLock) {
-      List<String> disabledReplicas = new ArrayList<>(getDisabledReplicas());
       String partitionName = replicaId.getPartitionId().toPathString();
 
       // 1. invoke Helix native method to enable/disable partition on local node, this will trigger subsequent state
       //    transition on given replica. This method modifies MapFields in InstanceConfig.
+      boolean instanceConfigUpdated = false;
       InstanceConfig instanceConfig = getInstanceConfig();
       String resourceNameForPartition = getResourceNameOfPartition(helixAdmin, clusterName, partitionName);
       logger.info("{} replica {} on current node", disable ? "Disabling" : "Enabling", partitionName);
-      instanceConfig.setInstanceEnabledForPartition(resourceNameForPartition, partitionName, !disable);
+      boolean currentlyEnabled = instanceConfig.getInstanceEnabledForPartition(resourceNameForPartition, partitionName);
+      if (currentlyEnabled == disable) {
+        instanceConfig.setInstanceEnabledForPartition(resourceNameForPartition, partitionName, !disable);
+        if (!helixAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig)) {
+          logger.warn("setReplicaDisabledState() failed InstanceConfig update");
+        }
+      }
```

Comment:
The current logic is correct but I am thinking if we can simplify this. `setReplicaDisabledState` method is only called when local node tries to enable/disable local replica. The enabled/disabled replica is always reflected by "disabled list" in dataNodeConfig. So I am thinking we don't have to check `currentlyEnabled == disable`. Instead, we may consider moving this piece of code into `if (dataNodeConfigChanged)` block. Only when the "disabled list" is updated, we call enable/disable partition in InstanceConfig. Hope my point makes some sense.

Reply:
Sure, that makes sense, and will save a HelixAdmin call in many cases. I will implement that